### ISSUE TITLE
Videoroom leaks 1 thread per data publisher in 1-to-many scenario

### ISSUE THREAD

**Issue by biodranik** [meetecho/janus-gateway#2367]
There is one "master" video/audio/data publisher, which subscribes to many data-only ("slave") publishers (no audio and video).
Each data-only publisher publishes only its data and subscribes only to the "master" publisher (to video/audio and data).

1. "Master" starts publishing it's media.
2. When every "slave" joins the room, total janus threads count increases by 3.
3. When any "slave" lefts, janus threads count decreases by 2 (1 thread out of 3 leaks). So for longer sessions, when many users join/left, janus has many leaked threads.
4. If "master" unpublishes without disconnecting, leaked threads are still there.
5. If "master" disconnects, leaked threads are destroyed properly.

Reproduces on the recent master on Mac and Linux, with H264/opus codecs.

**Comment by lminiero** [comment]
The VideoRoom plugin doesn't have thread. The threads you see are from the Janus core. By default, each handle (whether it has a PeerConnection active or not) has its own thread, since it needs one for the related event loop. As such, the threads will remain there until you detach the handles.

**Comment by atoppi** [comment]
> When any "slave" lefts, janus threads count decreases by 2 (1 thread out of 3 leaks). So for longer sessions, when many users join/left, janus has many leaked threads.

How is the slave leaving the room ? If you are just using `leave` API, bear in mind that Janus WILL NOT tear the RTC connection down and that may explain "leaking" threads and resources.
When leaving use `detach` to be totally sure that everything gets freed.

**Comment by biodranik** [comment]
It doesn't matter if the slave's browser tab/window is closed or handle.detach() or session.destroy() is called. The leak is still there, and it reproduces only with data channels enabled.
Is there anything special about datachannels? What else can I try to properly release the server's threads from the completely disconnected client (slave) side?

**Comment by atoppi** [comment]
I could not reproduce your issue doing these steps:

0. idle [`thread count = 24`]
1. publish a VR user (`master`) with audio+video+data [`thread count = 26`]
2. publish a VR user (`slave1`) with data-only
3. `master` subscribes to `slave1` data
4. `slave1` subscribes to `master` audio+video+data [`thread count = 29`]
5. repeat steps 2-4 for another `slave2` [`thread count = 32`]
6. detach `slave1` handles (both subscriber and publisher) [`thread count = 28`]
7. detach `slave2` handles (both subscriber and publisher) [`thread count = 25`]
8. detach `master` publisher handle
9. done [`thread count = 24`]

**Comment by lminiero** [comment]
> Is there anything special about datachannels?

I think usrsctp spawns an internal thread for every association. We get rid of the association when the PeerConnection is closed, though. You should check the name of the threads that remain "alive": we typically name ours, so they should be easy to spot.

**Comment by biodranik** [comment]
I found the issue: the subscriber (data-only) handle of the "master" did not call .detach() when "slave" publisher left the room/disconnected. Threads are freed correctly on the server side after explicitly calling subscriberHandle.detach().

I wrongly assumed that any subscriber handle should be automatically detached/destroyed if the publisher it subscribed to has left. Why is it not done automatically? How this "dangling" handle can be used after publisher has left?

**Comment by lminiero** [comment]
You should read the docs. Closing.

**Comment by biodranik** [comment]
VideoRoom docs don't say anything about it. My assumption was based on this Janus JS doc reference:
`detached: the plugin handle has been detached by the plugin itself, and so should not be used anymore.`

And I can't find anywhere, what is the use case of a subscriber handle when publisher has left. Why not to detach the handle by the plugin in this situation?

**Comment by atoppi** [comment]
Using a subscriber handle without a reference publisher is a totally valid design.
Indeed many applications exist that never attach a publisher handle and just use subscribers by exchanging streams id data outside of janus signaling.

**Comment by biodranik** [comment]
Ok, thanks for the explanation. That is indeed a valid use-case.

**Comment by lminiero** [comment]
> VideoRoom docs don't say anything about it

It does (look for `close_pc`).
